translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,355 @@
+		StartTimestamp: translatorcommon.UnixnanoToTimestamp(point.StartTimeUnixnano),
+		LabelValues:    labelsToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: translatorcommon.UnixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DoubleValue{
+					DoubleValue: point.Value,
+				},
+			},
+		},
+	}
+}
+
+func histogramPointToOC(point *otlpmetrics.HistogramDataPoint, labelKeys []*ocmetrics.LabelKey) *ocmetrics.TimeSeries {
+	return &ocmetrics.TimeSeries{
+		StartTimestamp: translatorcommon.UnixnanoToTimestamp(point.StartTimeUnixnano),
+		LabelValues:    labelsToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: translatorcommon.UnixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DistributionValue{
+					DistributionValue: &ocmetrics.DistributionValue{
+						Count:                 int64(point.Count),
+						Sum:                   point.Sum,
+						SumOfSquaredDeviation: 0,
+						BucketOptions:         histogramExplicitBoundsToOC(point.ExplicitBounds),
+						Buckets:               histogramBucketsToOC(point.Buckets),
+					},
+				},
+			},
+		},
+	}
+}
+
+func histogramExplicitBoundsToOC(bounds []float64) *ocmetrics.DistributionValue_BucketOptions {
+	if len(bounds) == 0 {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_BucketOptions{
+		Type: &ocmetrics.DistributionValue_BucketOptions_Explicit_{
+			Explicit: &ocmetrics.DistributionValue_BucketOptions_Explicit{
+				Bounds: bounds,
+			},
+		},
+	}
+}
+
+func histogramBucketsToOC(buckets []*otlpmetrics.HistogramDataPoint_Bucket) []*ocmetrics.DistributionValue_Bucket {
+	if len(buckets) == 0 {
+		return nil
+	}
+
+	ocBuckets := make([]*ocmetrics.DistributionValue_Bucket, 0, len(buckets))
+	for _, bucket := range buckets {
+		ocBuckets = append(ocBuckets, &ocmetrics.DistributionValue_Bucket{
+			Count:    int64(bucket.Count),
+			Exemplar: exemplarToOC(bucket.Exemplar),
+		})
+	}
+	return ocBuckets
+}
+
+func exemplarToOC(exemplar *otlpmetrics.HistogramDataPoint_Bucket_Exemplar) *ocmetrics.DistributionValue_Exemplar {
+	if exemplar == nil {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_Exemplar{
+		Value:       exemplar.Value,
+		Timestamp:   translatorcommon.UnixnanoToTimestamp(exemplar.TimestampUnixnano),
+		Attachments: exemplarAttachmentsToOC(exemplar.Attachments),
+	}
+}
+
+func exemplarAttachmentsToOC(attachments []*otlpcommon.StringKeyValue) map[string]string {
+	if len(attachments) == 0 {
+		return nil
+	}
+
+	ocAttachments := make(map[string]string, len(attachments))
+	for _, att := range attachments {
+		ocAttachments[att.Key] = att.Value
+	}
+	return ocAttachments
+}
+
+func labelsToOC(labels []*otlpcommon.StringKeyValue, labelKeys []*ocmetrics.LabelKey) []*ocmetrics.LabelValue {
+	if len(labels) == 0 {
+		return nil
+	}
+
+	// NOTE: We need to set label values in the same order as keys, thus
+	// intermediate transformation to a map for fast lookups
```

Comment:
Can you add a comment somewhere explaining what is done with the labels and keys here? I have a hard time following the logic which is scattered across many functions and multiple maps and slices.

Reply:
I left a few comments in label keys construction already, e.g.
```
	// Construct a list of label keys
	// Note: Label values will have to match keys by index
```
Let me know if it's worth adding more.

OTLP and OC have different representations of labels:
- OC has a single ordered list of label keys per metric in the MetricDescriptor; then, every data point has an ordered list of labels matching the keys
- In OTLP, every label is represented independently via `StringKeyValue`, i.e. theoretically different points may have different set of labels (not sure if that's allowed by the spec)

So what we do in this translator:
1) Scan all points and their labels to generate a unique set of all label keys used across the metric, sort them and set in MetricDescriptor
2) For each point we generate an ordered list of label values, matching the order of label keys passed here.

Reply:
Add this note to the code.